### PR TITLE
BCF header records

### PR DESCRIPTION
**Pull request by holtgrewe**
This works but I'm unsure about the design being right.

@johanneskoester what do you think?

### CODE REVIEW

File: src/bcf/header.rs [hunk 1]
```diff
@@ -240,6 +257,54 @@ impl HeaderView {
             (*(*self.inner).id[htslib::BCF_DT_SAMPLE as usize].offset(*id as isize)).key) };
         key.to_bytes().to_vec()
     }
+
+    /// Return structured `HeaderRecord`s.
+    pub fn header_records(&self) -> Vec<HeaderRecord> {
```
Comment (johanneskoester) [inline src/bcf/header.rs]: I would return a HashMap here, which assigns keys to HeaderRecord.

File: src/bcf/header.rs [hunk 2]
```diff
@@ -110,6 +110,23 @@ impl Drop for Header {
 }
 
 
+/// A header record.
+pub enum HeaderRecord {
+    /// A `FILTER` header record.
+    Filter { key: String, key_value_pairs: Vec<(String, String)> },
```
Comment (johanneskoester) [inline src/bcf/header.rs]: I think these should be `<type> {key: String, values: HashMap<String, String>}`, except for Generic.

File: src/bcf/header.rs (hunk 1 above)
Comment (johanneskoester) [reply src/bcf/header.rs]: We can use LinearMap instead, there are only a few keys involved here. See https://docs.rs/linear-map

File: src/bcf/header.rs (hunk 2 above)
Comment (johanneskoester) [reply src/bcf/header.rs]: Wait, this should be LinearMap.

File: src/bcf/header.rs (hunk 1 above)
Comment (johanneskoester) [reply src/bcf/header.rs]: Right. And moreover there might be duplicate keys I think. Returning a Vector is fine.

Comment (holtgrewe) [reply src/bcf/header.rs]: Hm, so should I just rebase or perform any changes?

Comment (holtgrewe) [reply src/bcf/header.rs]: OK, resolving the merge conflicts now FWIW, would have to be done anyway.

**Comment by holtgrewe** [comment]
🎉 @johanneskoester The test runs through, the code compiles and is properly formatted. Please advise on any further need for change.

**Comment by holtgrewe** [comment]
@johanneskoester ping

File: src/bcf/header.rs [hunk 3]
```diff
@@ -191,6 +191,38 @@ impl Drop for Header {
     }
 }
 
+/// A header record.
+#[derive(Debug)]
+pub enum HeaderRecord {
+    /// A `FILTER` header record.
+    Filter {
+        key: String,
+        key_value_pairs: Vec<(String, String)>,
```
[the remark this answers is not in the document]
Comment (holtgrewe) [reply src/bcf/header.rs]: OK, did so, let's see what the CI build says.

**Comment by holtgrewe** [comment]
@johanneskoester Well, this runs through in principle but doctests now also fail on stable... Otherwise, I incorporated your changes. What now?

File: Cargo.toml [hunk 4]
```diff
@@ -35,4 +36,4 @@ pretty_assertions = "0.5.1"
 
 [build-dependencies]
 fs-utils = "1.0"
-bindgen = "0.35.0"
+bindgen = "0.36.0"
```
Comment (johanneskoester) [inline Cargo.toml]: Without the trailing zero, just `0.36`? Then we automatically get patch releases.

**Comment by holtgrewe** [comment]
@johanneskoester No it's not. And now?

**Comment by johanneskoester** [comment]
This is either a bug in rust or a behavior change in rust and a bug in rust-bindgen. I have created an issue: https://github.com/rust-lang-nursery/rust-bindgen/issues/1313

**Comment by johanneskoester** [comment]
Can you update to the latest master of rust-htslib? I have temporarily pinned rust to 1.25

**Comment by johanneskoester** [comment]
Yay! Thanks a lot!

**Pull request merged**